enumerables.rb

```diff
@@ -0,0 +1,188 @@
+#!/home/manezeu/.rbenv/shims/ruby
+
+# rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Style/For, Metrics/BlockNesting, Layout/LineLength, Style/ConditionalAssignment, Metrics/ModuleLength, Metrics/AbcSize, Metrics/MethodLength, Style/NestedTernaryOperator
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    caller = self
+    if caller.class == Array || caller.class == Range
+      for item in caller
+        yield(item)
+      end
+    elsif caller.class == Hash
+      for item in caller.keys
+        yield(item, caller[item])
+      end
+    end
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    caller = self
+    if caller.class == Array
+      for index in (0...caller.length)
+        yield(caller[index], index)
+      end
+    elsif caller.class == Hash
+      for index in caller.keys
+        yield(caller[index], index)
+      end
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    caller = self
+    if caller.class == Array || caller.class == Range
+      arr = []
+      caller.my_each { |item| arr.push(item) if yield(item) }
+      arr
+    elsif caller.class == Hash
+      hash = {}
+      caller.my_each { |key, value| hash[key] = value if yield(key, value) }
+      hash
+    end
+  end
+
+  def my_all?(param = nil)
+    result = true
+    caller = self
+    if caller.class == Array || caller.class == Range
+      if !param.nil?
+        param.class == Regexp ? caller.my_each { |item| result = false unless param.match?(item) } : (param.class == String ? caller.my_each { |item| result = false unless item == param } : caller.my_each { |item| result = false unless item.class <= param })
+      else
+        block_given? ? caller.my_each { |item| result = false unless yield(item) } : caller.my_each { |item| result = false if !item || item.nil? }
+      end
+    elsif caller.class == Hash
+      if !param.nil?
+        param.class == Regexp ? caller.my_each { |_key, value| result = false unless my_pattern.match?(value) } : (param.class == String ? caller.my_each { |_key, value| result = false unless value == param } : caller.my_each { |_key, value| result = false unless value.class <= param })
+      else
+        block_given? ? caller.my_each { |_key, value| result = false unless yield(key, value) } : caller.my_each { |_key, value| result = false if !value || value.nil? }
+      end
+    end
+    result
```

Comment:
Extract some code into helper methods to reduce complexity, nesting, length, etc.